### PR TITLE
calculating remaining days until the server's certificate is expired

### PR DESCRIPTION
**Pull request by osherdp**
## What does this PR do?

This PR adds a data field in events of heartbeat, so that one can know how many days until the server's certificate is being expired.
In addition, a new set of tests has been added so that anyone can expend it for testing behavior relevant for SSL and heartbeat. If another test is of interest, let me know.

## Why is it important?

There are all sorts of data which is relevant to the subject: the expiration date, age, etc. All of them cannot tell us directly how close we are for the expiration of a certificate, thus not relevant unless we have a different logic to make the difference between expiration date and current time, such as scripted fields or Kibana's inner logic. ES query language cannot differentiate dates, scripted fields are less accessible for operators and not intuitive enough in general, and we might want to use different backends for alerting such as ElastAlert or Kibana's alerting system. That's why it's of value to just return the value and let anyone use it to watch for upcoming expirations.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by osherdp** [comment]
@elastic/uptime any thoughts about this issue?

**Comment by andrewvc** [comment]
Sorry for the delay, I'm not opposed to this approach, but I'd be curious as to where ES's query API is failing. ES does support dates, which we use in the Uptime UI.

My worry here is that this data is only useful really at the time of ingestion, the calculated values become wrong within 24 hours.

**Comment by andrewvc** [comment]
As an example of date queries you can do a search for:

```json
{"bool": {"filter": {"range": {"tls.server.not_before": {"lte": "now-30d"}}}}
```

**Comment by osherdp** [comment]
@andrewvc It's not wrong, as the data is relevant for the specific request that took place
Similarly, if a server return 5xx error code and the bug in the server is fixed afterward, it's OK that the saved data is relevant for the actual request & response that took place

Also, can you explain how is it possible via query language?
For example, how a user can do that in the "Discover" app?

**Comment by andrewvc** [comment]
@osherdp that's a good point that the data isn't wrong, it's definitely right at the time of ingestion. You're also write that it's hard to write a Kibana query in discover to match it. Given that, I'm reversing my opinion, I think this would be a useful addition.

**Comment by andrewvc** [comment]
I'll try to find some time to review it in the next few days.

File: go.mod [hunk 1]
```diff
@@ -123,7 +123,7 @@ require (
 	github.com/oklog/ulid v1.3.1
 	github.com/opencontainers/go-digest v1.0.0-rc1.0.20190228220655-ac19fd6e7483 // indirect
 	github.com/opencontainers/image-spec v1.0.2-0.20190823105129-775207bd45b6 // indirect
-	github.com/pierrre/gotestcover v0.0.0-20160113212533-7b94f124d338
+	github.com/pierrre/gotestcover v0.0.0-20160517101806-924dca7d15f0
```
Comment (andrewvc) [inline go.mod:126]: Can we undo this? Was there a specific reason this lib was bumped?

File: go.sum [hunk 2]
```diff
@@ -571,6 +571,8 @@ github.com/pierrec/lz4 v2.4.1+incompatible h1:mFe7ttWaflA46Mhqh+jUfjp2qTbPYxLB2/
 github.com/pierrec/lz4 v2.4.1+incompatible/go.mod h1:pdkljMzZIN41W+lC3N2tnIh5sFi+IEE17M5jbnwPHcY=
 github.com/pierrre/gotestcover v0.0.0-20160113212533-7b94f124d338 h1:/VAZ3an4jHXs+61iNHugNR1mG25MSpaxtMnwOJVEAQM=
 github.com/pierrre/gotestcover v0.0.0-20160113212533-7b94f124d338/go.mod h1:4xpMLz7RBWyB+ElzHu8Llua96TRCB3YwX+l5EP1wmHk=
+github.com/pierrre/gotestcover v0.0.0-20160517101806-924dca7d15f0 h1:i5VIxp6QB8oWZ8IkK8zrDgeT6ORGIUeiN+61iETwJbI=
```
Comment (andrewvc) [inline go.sum:574]: Let's not bump libs if it's not related

File: heartbeat/docs/fields.asciidoc [hunk 3]
```diff
@@ -7967,6 +7967,17 @@ example: 2019-08-16 01:40:25+00:00
 
 --
 
+*`tls.server.x509.days_to_expiry`*::
```
Comment (andrewvc) [inline heartbeat/docs/fields.asciidoc:7970]: Any reason to prefer days to millis? That should easily fit in a long and will probably be more useful.

**Review by andrewvc — CHANGES_REQUESTED** [review]
Apologies for the delay. This is looking really good, and I appreciate the extra effort on tests. I actually wouldn't have even said the python tests are necessary, but they are appreciated.

I left some minor feedback, and conflicts need to be resolved. Please ping me when those things are done!

File: heartbeat/docs/fields.asciidoc (hunk 3 above)
Comment (osherdp) [reply heartbeat/docs/fields.asciidoc]: Done

**Comment by osherdp** [comment]
@andrewvc all done
Also, elastic/ecs#979 should be pushed beforehand

**Comment by andrewvc** [comment]
jenkins, test this

**Comment by andrewvc** [comment]
Code looks great btw, let's see if CI is happy :)

**Comment by andrewvc** [comment]
jenkins, please test this

**Comment by andrewvc** [comment]
Hmmm, CI seems to not be triggering, I'll have someone take a look.

**Comment by mdelapenya** [comment]
jenkins run the tests please

**Comment by andrewvc** [comment]
jenkins run the tests please

**Review by andrewvc — APPROVED** [review]
LGTM

**Comment by andrewvc** [comment]
Thanks for being patient with this contribution! I'm going to target it for the 7.11 release.

**Review by andrewvc — CHANGES_REQUESTED** [review]
I was about to hit the merge button when I realized there is one small issue here, we need to add a definition for this new field to the mapping (it also shows up in the docs). Can you add one to: https://github.com/elastic/beats/blob/master/heartbeat/monitors/active/dialchain/_meta/fields.yml

Also, can you add an entry to https://github.com/elastic/beats/blob/master/CHANGELOG.next.asciidoc ?

We are almost there on this PR!

**Comment by osherdp** [comment]
@andrewvc according to this there's no need: https://github.com/elastic/ecs/pull/979#issuecomment-714629289

**Pull request closed without merging**